Index: common/advanced_config.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/common/advanced_config.cpp b/common/advanced_config.cpp
--- a/common/advanced_config.cpp
+++ b/common/advanced_config.cpp	(date 1773661661990)
@@ -72,6 +72,7 @@
 static const wxChar HoleWallThickness[] = wxT( "HoleWallPlatingThickness" );
 static const wxChar CoroutineStackSize[] = wxT( "CoroutineStackSize" );
 static const wxChar ShowRouterDebugGraphics[] = wxT( "ShowRouterDebugGraphics" );
+static const wxChar PNSExtraOptimizeOnRouting[] = wxT( "ExtraOptimizeOnRouting" );
 static const wxChar EnableRouterDump[] = wxT( "EnableRouterDump" );
 static const wxChar HyperZoom[] = wxT( "HyperZoom" );
 static const wxChar CompactFileSave[] = wxT( "CompactSave" );
@@ -323,6 +324,8 @@
 
     m_PNSProcessClusterTimeout = 100; // Default: 100 ms
 
+    m_ExtraOptimizeOnRouting = 0;
+
     loadFromConfigFile();
 }
 
@@ -437,6 +437,9 @@
     m_entries.push_back( std::make_unique<PARAM_CFG_BOOL>( true, AC_KEYS::ShowRouterDebugGraphics,
                                                            &m_ShowRouterDebugGraphics, m_ShowRouterDebugGraphics ) );

+    m_entries.push_back( std::make_unique<PARAM_CFG_INT>( 0, AC_KEYS::PNSExtraOptimizeOnRouting,
+                                                           &m_ExtraOptimizeOnRouting, m_ExtraOptimizeOnRouting, 0, 256 ) );
+
     m_entries.push_back( std::make_unique<PARAM_CFG_BOOL>( true, AC_KEYS::EnableRouterDump, &m_EnableRouterDump,
                                                            m_EnableRouterDump ) );
 
Index: pcbnew/router/pns_line_placer.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/pcbnew/router/pns_line_placer.cpp b/pcbnew/router/pns_line_placer.cpp
--- a/pcbnew/router/pns_line_placer.cpp
+++ b/pcbnew/router/pns_line_placer.cpp	(date 1773661661993)
@@ -25,6 +25,8 @@
 #include "pns_arc.h"
 #include "pns_debug_decorator.h"
 #include "pns_line_placer.h"
+
+#include "advanced_config.h"
 #include "pns_node.h"
 #include "pns_router.h"
 #include "pns_shove.h"
@@ -1952,7 +1954,7 @@
     LINE l_orig = aNode->AssembleLine( aLatest, nullptr, false, false, false );
     LINE l( l_orig );
 
-    bool optimized = OPTIMIZER::Optimize( &l, OPTIMIZER::MERGE_COLINEAR, aNode );
+    bool optimized = OPTIMIZER::Optimize( &l, OPTIMIZER::MERGE_COLINEAR | ADVANCED_CFG::GetCfg().m_ExtraOptimizeOnRouting, aNode );
 
     SHAPE_LINE_CHAIN simplified( l.CLine() );
 
Index: include/advanced_config.h
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/include/advanced_config.h b/include/advanced_config.h
--- a/include/advanced_config.h
+++ b/include/advanced_config.h	(date 1773661693288)
@@ -241,6 +241,15 @@
      */
     bool m_ShowRouterDebugGraphics;
 
+    /**
+     * Extra optimizations on routing
+     *
+     * Setting name: "ExtraOptimizeOnRouting"
+     * Valid values: 0 to 256
+     * Default value: 0
+     */
+    int m_ExtraOptimizeOnRouting;
+
     /**
      * Enable PNS router to dump state information for debug purpose (press `0` while routing)
      *
