diff --git a/lib/libf2fs.c b/lib/libf2fs.c
index 93b0604..805a992 100644
--- a/lib/libf2fs.c
+++ b/lib/libf2fs.c
@@ -959,17 +959,28 @@ int get_device_info(int i)
 		}
 	}
 
+	// force the desired sector size for all devices by default
+	if (c.wanted_sector_size != -1) {
+		dev->sector_size = c.wanted_sector_size;
+	}
+
 	if (c.sparse_mode) {
 		dev->total_sectors = c.device_size / dev->sector_size;
 	} else if (S_ISREG(stat_buf->st_mode)) {
 		dev->total_sectors = stat_buf->st_size / dev->sector_size;
 	} else if (S_ISBLK(stat_buf->st_mode)) {
+		// force desired sector size for a block device
+		if (c.wanted_sector_size != -1) {
+			dev->sector_size = c.wanted_sector_size;
+		}
+		else {
 #ifdef BLKSSZGET
 		if (ioctl(fd, BLKSSZGET, &sector_size) < 0)
 			MSG(0, "\tError: Using the default sector size\n");
 		else if (dev->sector_size < sector_size)
 			dev->sector_size = sector_size;
 #endif
+		}
 #ifdef BLKGETSIZE64
 		if (ioctl(fd, BLKGETSIZE64, &dev->total_sectors) < 0) {
 			MSG(0, "\tError: Cannot get the device size\n");
